tidy3d/plugins/mode/solver.py

```diff
@@ -409,5 +454,84 @@ def solver_eigs(mat, num_modes, guess_value=1.0):
     guess_value : float, optional
     """
 
-    values, vectors = spl.eigs(mat, k=num_modes, sigma=guess_value, tol=fp_eps)
+    values, vectors = spl.eigs(mat, k=num_modes, sigma=guess_value, tol=fp_eps, v0=vec_init)
     return values, vectors
+
+
+def isinstance_complex(vec_or_mat, tol=TOL_COMPLEX):
+    """Check if a numpy array or scipy csr_matrix has complex component by looking at
+    norm(x.imag)/norm(x)>TOL_COMPLEX
+
+    Parameters
+    ----------
+    vec_or_mat : Union[np.ndarray, sp.csr_matrix]
+    """
+
+    if isinstance(vec_or_mat, np.ndarray):
+        return np.linalg.norm(vec_or_mat.imag) / np.linalg.norm(vec_or_mat) > tol
+    if isinstance(vec_or_mat, sp.csr_matrix):
+        return spl.norm(vec_or_mat.imag) / spl.norm(vec_or_mat) > tol
+
+    raise RuntimeError("Variable type should be either numpy array or scipy csr_matrix.")
+
+
+def type_conversion(vec_or_mat, new_dtype):
+    """Convert vec_or_mat to new_type.
+
+    Parameters
+    ----------
+    vec_or_mat : Union[np.ndarray, sp.csr_matrix]
+        vec or mat to be converted.
+    new_dtype : Union[np.complex128, np.complex64, np.float64, np.float32]
+        Final type of vec or mat
+
+    Returns
+    -------
+    converted_vec_or_mat : Union[np.ndarray, sp.csr_matrix]
+    """
+
+    if new_dtype in {np.complex128, np.complex64}:
+        return vec_or_mat.astype(new_dtype)
+    if new_dtype in {np.float64, np.float32}:
+        converted_vec_or_mat = vec_or_mat.real
+        return converted_vec_or_mat.astype(new_dtype)
+
+    raise RuntimeError("Unsupported new_type.")
+
+
+def set_initial_vec(Nx, Ny, mat_dtype=np.complex128, is_tensorial=False):
+    """Set initial vector for eigs:
+    1) The field at x=0 and y=0 boundaries are set to 0. This should be
+    the case for PEC boundaries, but wouldn't hurt for non-PEC boundary;
+    2) The vector is np.complex128 by default, and will be converted to
+    approriate type afterwards.
+
+    Parameters
+    ----------
+    Nx : int
+        Number of grids along x-direction.
+    Ny : int
+        Number of grids along y-direction.
+    mat_dtype : Union[np.complex128, np.complex64, np.float64, np.float32]
+        The type of returned vector.
+    is_tensorial : bool
+        diagonal or tensorial eigenvalue problem.
+    """
+
+    # The size of the vector is len_multiplier * Nx * Ny
+    len_multiplier = 2
+    if is_tensorial:
+        len_multiplier *= 2
+
+    # Initialize the vector
+    size = (Nx, Ny, len_multiplier)
+    vec_init = np.random.random(size) + 1j * np.random.random(size)
+    vec_init = type_conversion(vec_init, mat_dtype)
+
+    # Set values at the boundary to be 0
+    vec_init[0, :, :] = 0
+    vec_init[:, 0, :] = 0
```

Comment:
These need to happen only if Nx > 1 and Ny > 1, correspondingly. For a 1D "plane", the derivative matrices impose a periodic boundary condition along that direction.

Reply:
see the most recent commit for addressing your two comments.